### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by northys** [krateng/maloja#132]
I've added maloja to my uptime monitoring to check home page every minute and it killed my homelab completely.

```
Apr 24 09:49:59 nuc.northys.cz kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=d922e8d4fcf01a86bd2d4>
Apr 24 09:49:59 nuc.northys.cz kernel: Out of memory: Killed process 758402 (Maloja) total-vm:7568084kB, anon-rss:75306>
Apr 24 09:49:59 nuc.northys.cz kernel: oom_reaper: reaped process 758402 (Maloja), now anon-rss:0kB, file-rss:0kB, shme>
```

**Comment by northys** [comment]
```
root@nuc ~ # journalctl --since "12 hours ago" | grep -i maloja
Apr 24 01:48:16 nuc.northys.cz kernel: [ 740653]     0 740653  2182572  2172736 18059264        0             0 Maloja
Apr 24 01:48:16 nuc.northys.cz kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=b97b7b6ba2230a8da6ff290af2fbe2c4aa566146c5564df9fbc79da815e194ed,mems_allowed=0,global_oom,task_memcg=/docker/ebc7ae2133b2684874f1f4036f4cb537b43291a5aa46c61de1e2f2b2adc02598,task=Maloja,pid=740653,uid=0
Apr 24 01:48:16 nuc.northys.cz kernel: Out of memory: Killed process 740653 (Maloja) total-vm:8730288kB, anon-rss:8690944kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:17636kB oom_score_adj:0
Apr 24 01:48:16 nuc.northys.cz kernel: oom_reaper: reaped process 740653 (Maloja), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
Apr 24 09:49:59 nuc.northys.cz kernel: [ 758402]     0 758402  1892021  1882657 15622144        0             0 Maloja
Apr 24 09:49:59 nuc.northys.cz kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=d922e8d4fcf01a86bd2d4062169a5a31ef69478b6eb9df7258cc348bb4fd2f4e,mems_allowed=0,global_oom,task_memcg=/docker/ebc7ae2133b2684874f1f4036f4cb537b43291a5aa46c61de1e2f2b2adc02598,task=Maloja,pid=758402,uid=0
Apr 24 09:49:59 nuc.northys.cz kernel: Out of memory: Killed process 758402 (Maloja) total-vm:7568084kB, anon-rss:7530628kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:15256kB oom_score_adj:0
Apr 24 09:49:59 nuc.northys.cz kernel: oom_reaper: reaped process 758402 (Maloja), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
root@nuc ~ #
```

**Comment by krateng** [comment]
Hm, is that with cache enabled?

While this certainly doesn't fix the issue you could replace the homepage check with a call to `/apis/mlj_1/serverinfo` for now which at least avoids generating all the stats every minute.

**Comment by northys** [comment]
Yup it's with cache enabled. I'll make coffee and dig more into syslog, graphs I have in prometheus and into logs produced by maloja.

**Comment by northys** [comment]
```

**Comment by northys** [comment]
Looks like it took only 2 hours to consume about 8GB of RAM.

![image](https://user-images.githubusercontent.com/2878126/164980474-3f7c6c3e-a352-417b-b262-90771399d22c.png)

**Comment by krateng** [comment]
Alright, have reproduced it, now working on profiling memory use... this will take some time.

**Comment by northys** [comment]
I wonder whether the cache is used later? Because I haven't scrobbled a single song during the night and my sqlite database is only 30MB big. 

I understand that it is CPU consuming to calculate all those SUMs but now it feels like the HP cache is created every request instead of page being server instantly because stats are already in cache. So it may be "same results are being cached over and over again" instead of "cache memory leak".

What do you think?

**Comment by krateng** [comment]
I can only guess that when memory usage gets too high, the cache is expelled, but somehow some references remain, meaning the OS can't reclaim that memory. Then, on the next generation of the homepage, the cache will be filled again with the same data because the actual cache doesn't reference it anymore - so it's indeed the same data over and over again, just not referenced properly. I'm probably gonna disable the cache per default for now until I have more time to profile this properly.

The cache is using the `lru-dict` library with the caching logic implemented myself.

**Comment by krateng** [comment]
Alright, maybe I fixed it. Off to more testing! 😅

**Comment by krateng** [comment]
Okay, I've pushed a new version. The homepage should now no longer cause RAM usage to go out of control. It still doesn't really explain why the cache expulsion didn't trigger, but at least pinging your homepage should now be safe.

**Comment by northys** [comment]
> It still doesn't really explain why the cache expulsion didn't trigger

Has that happened also for you?

I could explain that to myself as my homepab has onl, 12GB of memory and running a lot of stuff. E.g. every hour it runs 20 rsync and one restic.io backup which loads some backup repository index into memory. This index may have like 2GB which causes system to go into OOM state and OOM killer took maloja down before it could realize that it may clear cache. Wouldn't that make sence?

**Comment by northys** [comment]
Since maloja is single user application I would set some hardcoded limit for cache. My database has 50MB at most so I would expect that 512MB for cache si just too much. I would set the cache limit to 256MB or maybe even 128MB.

**Comment by krateng** [comment]
The lru_dict structure doesn't support that unfortunately (and it's also somewhat hard to estimate the memory use of an object that consists of pointers), so I have to actively clear the cache depending on memory use. The entries can be quite big compared to the raw DB size, so there's definitely a performance benefit of having 512MB or more - and if the RAM is available on the system, it should be used after all, as lots of empty RAM doesn't do anything. I've settled for now on cache clearance at 40% total RAM use (overwritable by user setting) plus some extra log output - let's see how it performs.

**Comment by northys** [comment]
I've just upgraded to latest version and homepage is much faster. Good job!

> as lots of empty RAM doesn't do anything

That's true but I would be more happy if system could use that free ram for filesystem cache rather than to waste it by a program which uses half of my RAM to cache some data I access twise a week.

> overwritable by user setting

great!
